ts/Base/MaterialComponent.ts

```diff
@@ -0,0 +1,146 @@
+ * @class MaterialComponent
+ * @extends {Component}
+ */
+export abstract class MaterialComponent<
+  PropType extends {[prop: string]: any},
+  StateType extends {[prop: string]: any}
+> extends Component<
+  MaterialComponentProps<PropType>,
+  MaterialComponentState<StateType>
+> {
+  /**
+   * Attributes inside this array will be check for boolean value true
+   * and will be converted to mdc classes
+   */
+  protected abstract mdcProps: string[];
+  /** This will again be used to add apt classname to the component */
+  protected abstract componentName: string;
+
+  /** The final class name given to the dom */
+  protected classText?: string | null;
+  protected ripple?: MDCRipple | null;
+
+  protected control?: Element;
+
+  public render(props): VNode {
+    if (!this.classText) {
+      this.classText = this.buildClassName();
+    }
+    // Fetch a VNode
+    const componentProps = props;
+    const userDefinedClasses =
+      componentProps.className || componentProps.class || '';
+
+    // We delete class props and add them later in the final
+    // step so every component does not need to handle user specified classes.
+    if (componentProps.class) {
+      delete componentProps.class;
+    }
+    if (componentProps.className) {
+      delete componentProps.className;
+    }
+
+    const element = this.materialDom(componentProps);
+    element.attributes = element.attributes || {};
+
+    element.attributes.className = `${userDefinedClasses} ${this.getClassName(
+      element
+    )}`
+      .split(' ')
+      .filter(
+        (value, index, self) => self.indexOf(value) === index && value !== ''
+      ) // Unique + exclude empty class names
+      .join(' ');
+    // Clean this shit of proxy attributes
+    this.mdcProps.forEach(prop => {
+      delete element.attributes[prop];
+    });
+    return element;
+  }
+
+  /** Attach the ripple effect */
+  public componentDidMount() {
+    if (this.props.ripple && this.control) {
+      this.ripple = new MDCRipple(this.control);
+    }
+  }
+
+  public componentWillUnmount() {
+    if (this.ripple) {
+      this.ripple.destroy();
+    }
+  }
+
+  // Shared setter for the root element ref
+  @autobind
+  protected setControlRef(control: Element | undefined) {
+    this.control = control;
+  }
+
+  /** Build the className based on component names and mdc props */
+  @autobind
+  protected buildClassName() {
+    // Class name based on component name
+    let classText = 'mdc-' + this.componentName;
+
+    // Loop over mdcProps to turn them into classNames
+    for (const propKey in this.props) {
+      if (this.props.hasOwnProperty(propKey)) {
+        const prop = this.props[propKey];
+        if (typeof prop === 'boolean' && prop) {
+          if (this.mdcProps.indexOf(propKey) !== -1) {
+            classText += ` mdc-${this.componentName}--${propKey}`;
+          }
+        }
+      }
+    }
+
+    return classText;
+  }
+
+  /** Returns the class name for element */
+  @autobind
```

Comment:
nit: @cromefire @YePpHa can you guys just check if this is really required?
I see no reason why this decorator is needed

Reply:
Yes removing it results in `TypeError: this is undefined`

Reply:
it is binding `this` to the class object